Signal/src/ViewControllers/ConversationView/ConversationViewModel.m

```diff
@@ -0,0 +1,1284 @@
+}
+
+- (YapDatabaseConnection *)editingDatabaseConnection
+{
+    return self.primaryStorage.dbReadWriteConnection;
+}
+
+- (OWSContactsManager *)contactsManager
+{
+    return (OWSContactsManager *)SSKEnvironment.shared.contactsManager;
+}
+
+- (OWSBlockingManager *)blockingManager
+{
+    return OWSBlockingManager.sharedManager;
+}
+
+#pragma mark
+
+- (void)addNotificationListeners
+{
+    [[NSNotificationCenter defaultCenter] addObserver:self
+                                             selector:@selector(uiDatabaseDidUpdateExternally:)
+                                                 name:OWSUIDatabaseConnectionDidUpdateExternallyNotification
+                                               object:self.primaryStorage.dbNotificationObject];
+    [[NSNotificationCenter defaultCenter] addObserver:self
+                                             selector:@selector(uiDatabaseWillUpdate:)
+                                                 name:OWSUIDatabaseConnectionWillUpdateNotification
+                                               object:self.primaryStorage.dbNotificationObject];
+    [[NSNotificationCenter defaultCenter] addObserver:self
+                                             selector:@selector(uiDatabaseDidUpdate:)
+                                                 name:OWSUIDatabaseConnectionDidUpdateNotification
+                                               object:self.primaryStorage.dbNotificationObject];
+    [[NSNotificationCenter defaultCenter] addObserver:self
+                                             selector:@selector(applicationDidEnterBackground:)
+                                                 name:OWSApplicationDidEnterBackgroundNotification
+                                               object:nil];
+    [[NSNotificationCenter defaultCenter] addObserver:self
+                                             selector:@selector(signalAccountsDidChange:)
+                                                 name:OWSContactsManagerSignalAccountsDidChangeNotification
+                                               object:nil];
+}
+
+- (void)signalAccountsDidChange:(NSNotification *)notification
+{
+    OWSAssertIsOnMainThread();
+
+    [self ensureDynamicInteractions];
+}
+
+- (void)configure
+{
+    OWSLogInfo(@"");
+
+    // We need to update the "unread indicator" _before_ we determine the initial range
+    // size, since it depends on where the unread indicator is placed.
+    self.lastRangeLength = 0;
+
+    [self ensureDynamicInteractions];
+    [self.primaryStorage updateUIDatabaseConnectionToLatest];
+
+    [self createNewMessageMappings];
+    if (![self reloadViewItems]) {
+        OWSFailDebug(@"failed to reload view items in configureForThread.");
+    }
+}
+
+- (void)viewDidLoad
+{
+    [self addNotificationListeners];
+}
+
+- (void)dealloc
+{
+    [[NSNotificationCenter defaultCenter] removeObserver:self];
+}
+
+- (BOOL)canLoadMoreItems
+{
+    if (self.lastRangeLength >= kYapDatabaseRangeMaxLength) {
```

Comment:
The vast majority of this code in the CVM was drawn from the CVC unchanged.  I'll try to call out big differences.